SEED-platform/seed
# Create property timeline view

#### Any background context you want to provide?
See #3769

#### What's this PR do?
- Adds timeline features 
- Refactors scenario and measure layout
- Migration to retroactively create events for existing databases
- Auto generates events when a Note, Analysis, or Building File is created
- Notes and Analyses have a cascade delete relationship with events 


#### How should this be manually tested?
setup:
- populate a single property across multiple cycles 

Notes:
- create a note in cycle A, create a note in cycle B. Check the timeline view and ensure notes are in the correct cycles
- delete a note and ensure the timeline event has also been deleted 

Analyses:
- create an analysis in cycle A and in cycle B. Check the timeline for the correct placement 
- delete an analysis and ensure the event has also been deleted 

AT Uploads:
- upload 2 building sync files for cycle A. upload a building sync file for Cycle B. Ensure the scenarios and measures table at the bottom of the property detail shows the correct AT files for the selected cycle.
- Check the timeline for ATEvents 
- Go to property detail and delete a scenario. Check the timeline has been updated for the removed scenario.

Timeline:
- toggle "collapse all" and "expand all". Confirm expected functionality 
- sort by "Date, Oldest to Newest", sort by "Date, Newest to Oldest"
- Use the event selection to deselect a type of event. Ensure the type is not listed in the timeline.
 

#### What are the relevant tickets?
#3769 Create Property Timeline Table
#3843 Create property timeline view UI
#3768 Create a new navigation item in property detail for Property Timeline
#3890 Edit Scenarios and measures on Property Details Page
#3777 Add a view set for measures

#### Screenshots (if appropriate)
[Mockups](https://xd.adobe.com/view/9516edb6-76a6-4a19-8b74-203c7f32385e-c045/grid/)

Property Detail Scenarios and Measures
![Screenshot 2023-03-20 at 11 35 56 AM](https://user-images.githubusercontent.com/58446472/226424004-8060b0ac-3624-4eb7-89cf-067300559b78.png)

Timeline Collapsed
![Screenshot 2023-03-20 at 11 33 12 AM](https://user-images.githubusercontent.com/58446472/226424062-1a0bb755-f0cd-4abb-b73d-953d53c5e0d4.png)


Timeline Expanded
![Screenshot 2023-03-20 at 11 34 42 AM](https://user-images.githubusercontent.com/58446472/226424089-9acf8715-a4f3-405a-9269-1534a8f29bda.png)

Event Selection
![Screenshot 2023-03-20 at 11 35 15 AM](https://user-images.githubusercontent.com/58446472/226424448-bd92c4fe-4b1a-4560-a887-c6e7327c4688.png)
